### ISSUE TITLE
docs unclear on how to calculate distance matrix for 1 single frame

### ISSUE THREAD

**Issue by bionicles** [mdtraj/mdtraj#1457]
i want to calculate a distance matrix for one single step, instead of a whole trajectory. is this possible with mdtraj? i guess i could just wrap the single step in an array and make a trajectory of 1. would it be possible to improve the docs to have details about how to calculate interatomic distances for 1 step?

**Comment by justinrporter** [comment]
Have you tried just slicing into the trajectory?

```python
import mdtraj as md
trj = md.load('my_trajectory.h5')
dists = md.compute_distances(trj[0], atom_pairs=np.arange(trj.top.n_atoms))
```

Then, `dists` will be an n x n x 1 all-to-all distance matrix.

**Comment by bionicles** [comment]
Thank you. I was curious to just pass a single step to some function and get back the distances for a loss function. This solution requires us to make a trajectory, pass it to a function, and slice it; I will continue to do it with tensorflow. It would be cool to have a way to pass 1 tensor or NumPy array of Num_atoms, XYZ to a mdtraj function and get back the distance tensors in same format they went in, for deep learning it’s cool if we can preserve same type and shape... fewer bugs

One option might be to switch to off policy RL algorithm and pass a trajectory of tensors to the function during the update step

**Comment by bionicles** [comment]
Maybe something like

def meta_distance_loss(initial, trajectory):
    initial_distances = cdist(initial)
    return sum([ mse(initial_distances, cdist(xyz)) for xyz in traj ])

def loss(batch):
    return sum([ meta_distance_loss(initial, traj) for initial, traj in batch])

Not sure if this makes sense. I want to get the sum of mean squared error between initial distances and the distances for each step in a trajectory as quickly as possible, but my xyz positions see in the form of tensorflow tensors

**Comment by bionicles** [comment]
Probably wise to normalize this value by num atoms squared or something like that

**Comment by bionicles** [comment]
@justinrporter 
how do you get the xyz vector each atom needs to move?

**Comment by justinrporter** [comment]
> I was curious to just pass a single step to some function and get back the distances for a loss function.

Do you not need differentiability? MDTraj has implemented its RMSD in C and so getting at the derivatives might be hard. If they even exist--I think there is an iterative component to the QCP algorithm that's used.

> Sometimes in bigger proteins the GPU memory is exceeded and the system will crash. Can MDTraj handle bigger proteins than fit in memory?

Are you talking about system RAM or GPU RAM here? Do you really have a protein that is so big it exceeds RAM? The largest known protein is titin, of 30k residues--call it 200k heavy atoms or so... 200k * 3 dimensions * 32 bits is only 2GB, right?

That being said, you might be able to replace the `xyz` parameter of the `Trajectory` with a numpy memory map. I haven't tried this but it might work.

>  how do you get the xyz vector each atom needs to move?

If all you want is the best superposition of a protein, you can make a copy, use `Trajectory.superpose` on the copy, and subtract the `xyz` parameters of the two. That'll be differentiable but much slower than `md.rmsd`. MDTraj's `md.rmsd` implementation actually avoids computing the superposition using [Theobald's QCP method](https://theobald.brandeis.edu/qcp/).

Another option is to bail on the 3d coordinates entirely and use the matrix of all pairwise distances. This will get out of control very quickly though, since it grows with the square of the size of your protein.
